File: backend/app/services/weather_service.py
import os
import aiohttp
from typing import Dict, Any
from dotenv import load_dotenv

load_dotenv()

class WeatherService:
    def __init__(self):
        self.api_key = os.getenv("OPENWEATHER_API_KEY")
        self.base_url = "https://api.openweathermap.org/data/2.5"

    async def get_weather(self, lat: float, lon: float) -> Dict[str, Any]:
        """获取特定位置的天气数据"""
        url = f"{self.base_url}/weather"
        params = {
            "lat": lat,
            "lon": lon,
            "appid": self.api_key,
            "units": "metric"
        }
        
        async with aiohttp.ClientSession() as session:
            async with session.get(url, params=params) as response:
                if response.status == 200:
                    return await response.json()
                return {}

    async def get_weather_forecast(self, lat: float, lon: float) -> Dict[str, Any]:
        """获取天气预报数据"""
        url = f"{self.base_url}/forecast"
        params = {
            "lat": lat,
            "lon": lon,
            "appid": self.api_key,
            "units": "metric"
        }
        
        async with aiohttp.ClientSession() as session:
            async with session.get(url, params=params) as response:
                if response.status == 200:
                    return await response.json()
                return {}

    def calculate_weather_impact(self, weather_data: Dict[str, Any]) -> float:
        """
        计算天气对电动车续航的影响
        返回一个影响系数（0-1之间，1表示最大影响）
        """
        if not weather_data:
            return 0.0

        impact = 0.0
        temp = weather_data.get("main", {}).get("temp", 20)
        humidity = weather_data.get("main", {}).get("humidity", 50)
        wind_speed = weather_data.get("wind", {}).get("speed", 0)

        # 温度影响
        if temp < 0:
            impact += 0.3
        elif temp < 10:
            impact += 0.2
        elif temp > 30:
            impact += 0.15

        # 湿度影响
        if humidity > 80:
            impact += 0.1

        # 风速影响
        if wind_speed > 20:
            impact += 0.2
        elif wind_speed > 10:
            impact += 0.1

        return min(impact, 1.0) 

File: hci_final_navigation/WeatherService.swift
import Foundation
import CoreLocation

struct WeatherInfo: Codable {
    let temperature: Double
    let condition: String
    let humidity: Int
    let windSpeed: Double
    let precipitation: Double
}

class WeatherService {
    private let apiKey: String
    
    init(apiKey: String) {
        self.apiKey = apiKey
    }
    
    func getWeather(for location: CLLocationCoordinate2D) async throws -> WeatherInfo {
        let urlString = "https://api.openweathermap.org/data/2.5/weather?lat=\(location.latitude)&lon=\(location.longitude)&appid=\(apiKey)&units=metric"
        
        guard let url = URL(string: urlString) else {
            throw URLError(.badURL)
        }
        
        let (data, _) = try await URLSession.shared.data(from: url)
        let response = try JSONDecoder().decode(WeatherResponse.self, from: data)
        
        return WeatherInfo(
            temperature: response.main.temp,
            condition: response.weather.first?.main ?? "",
            humidity: response.main.humidity,
            windSpeed: response.wind.speed,
            precipitation: response.rain?.oneHour ?? 0
        )
    }
    
    func getTravelAdvice(weather: WeatherInfo) -> String {
        var advice: [String] = []
        
        // 温度建议
        if weather.temperature > 30 {
            advice.append("高温天气，请注意车内温度调节")
        } else if weather.temperature < 5 {
            advice.append("低温天气，电池效率可能降低，建议提前规划充电")
        }
        
        // 降水建议
        if weather.precipitation > 0 {
            advice.append("有降水，路面可能湿滑，请谨慎驾驶")
        }
        
        // 风速建议
        if weather.windSpeed > 20 {
            advice.append("强风天气，可能影响行驶稳定性")
        }
        
        return advice.isEmpty ? "天气适宜出行" : advice.joined(separator: "。")
    }
}

// API 响应模型
private struct WeatherResponse: Codable {
    let weather: [Weather]
    let main: Main
    let wind: Wind
    let rain: Rain?
}

private struct Weather: Codable {
    let main: String
    let description: String
}

private struct Main: Codable {
    let temp: Double
    let humidity: Int
}

private struct Wind: Codable {
    let speed: Double
}

private struct Rain: Codable {
    let oneHour: Double?
    
    enum CodingKeys: String, CodingKey {
        case oneHour = "1h"
    }
} 